example_envs/bookinfo/productpage.rs

```diff
@@ -24,54 +24,58 @@ impl fmt::Display for ProductPage {
 impl SimElement for ProductPage {
     fn tick(&mut self, tick: u64) -> Vec<(Rpc, String)> {
         let mut ret = vec![];
-        let review_nodes = vec!["reviews-v1", "reviews-v2", "reviews-v3"];
-        let mut rng: StdRng = SeedableRng::seed_from_u64(self.core_node.seed);
-
         for _ in 0..min(
             self.core_node.queue.size(),
             self.core_node.egress_rate as usize,
         ) {
-            let mut rpc: Rpc;
+            let rpc: Rpc;
             if self.core_node.queue.size() > 0 {
                 let deq = self.core_node.dequeue(tick);
                 rpc = deq.unwrap();
             } else {
-                // No RPC in the queue and we only forward. So nothing to do
+                // no rpc in the queue, we only forward so nothing to do
                 continue;
             }
-            // Forward requests/responses from the gateway to one of reviews
-            // Also send a request to details-v1
-            // If we get a reply from one of reviews or details,
-            // send it back to the gateway
+            // forward requests/responses from productpage or reviews
             if rpc.headers.contains_key("src") {
-                let source: &str = &rpc.headers["src"];
-                if review_nodes.contains(&source) || source == "details-v1" {
-                    rpc.headers
-                        .insert("src".to_string(), self.core_node.id.to_string());
-                    ret.push((rpc, "gateway".to_string()));
-                } else if source == "gateway" {
-                    let idx = rng.gen_range(0, review_nodes.len());
-                    let dest = review_nodes[idx];
-                    rpc.headers
+                let destinations = self.choose_destination(&rpc);
+                for dest in destinations {
+                    let mut new_rpc = rpc.clone();
+                    new_rpc.headers.insert("dest".to_string(), dest.clone());
+                    new_rpc
+                        .headers
                         .insert("src".to_string(), self.core_node.id.to_string());
-                    ret.push((rpc.clone(), dest.to_string()));
-                    // also send a request to details-v1
-                    ret.push((rpc, "details-v1".to_string()));
-                } else if source == &self.core_node.id {
-                    // if we are creating a new RPC, eg, we are sending to storage
-                    let dest: &str = &rpc.headers["dest"];
-                    ret.push((rpc.clone(), dest.to_string()));
-                } else {
-                    panic!("ProductPage node does not have a valid source!");
+                    new_rpc
+                        .headers
+                        .insert("location".to_string(), "egress".to_string());
+                    if self.core_node.plugin.is_some() {
+                        self.core_node.plugin.as_mut().unwrap().recv(
```

Comment:
This is incorrect. We do not want to call recv here.

Reply:
We do want to call recv.  Otherwise the plugin will not receive the rpc as it leaves the node.

Reply:
Whoops, you are correct, my bad.